### PR TITLE
Issuance.sol

### PR DESCRIPTION
**Pull request by uivlis**
Note: need to add tests.

### CODE REVIEW

**Comment by uivlis** [comment]
Alright, ready for review.

File: contracts/drafts/issuance/Issuance.sol [hunk 1]
```diff
@@ -113,6 +116,34 @@ contract Issuance is Ownable, StateMachine {
         emit InvestmentAdded(msg.sender, _amount);
     }
 
+    function withdraw() external nonReentrant {
+        require(currentState == "DISTRIBUTING", "Cannot withdraw now.");
```
Comment (albertocuestacanada) [inline contracts/drafts/issuance/Issuance.sol]: You probably don't need a DISTRIBUTING state and can move directly from OPEN to LIVE, allowing investors to withdraw if we are LIVE.

File: contracts/drafts/issuance/Issuance.sol [hunk 2]
```diff
@@ -113,6 +116,34 @@ contract Issuance is Ownable, StateMachine {
...
+        require(investments[msg.sender] > 0, "No investments found.");
+        if (nextInvestor < investors.length) {
```
Comment (albertocuestacanada) [inline contracts/drafts/issuance/Issuance.sol]: You probably don't need `nextInvestor` anymore. Investors can come and serve themselves in any order.

File: contracts/drafts/issuance/Issuance.sol [hunk 3]
```diff
@@ -146,49 +177,37 @@ contract Issuance is Ownable, StateMachine {
...
 
     /**
-     * @dev Function to cancel all investments and close the issuance
+     * @dev Function to call repeatedly from frontend, until all investors are refunded
      */
-    function cancelAllInvestments() public onlyOwner{
-        require (
-            currentState == "OPEN",
-            "Cannot cancel now."
+    function refundNextInvestor() public onlyOwner nonReentrant {
```
Comment (albertocuestacanada) [inline contracts/drafts/issuance/Issuance.sol]: You can do the same as for `withdraw`. If the issuance goes to a FAILED state then you can just let every investor come and execute a `refund` function that gives them their money back.

File: contracts/drafts/issuance/Issuance.sol [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            "Not enough funds collected."
+        );
+        transition("LIVE");
+    }
+
+    /**
+     * @dev Function to cancel all investments
+     */
+    function cancelAllInvestments() public onlyOwner{
+        require (
```
Comment (albertocuestacanada) [inline contracts/drafts/issuance/Issuance.sol]: `transition("FAILED")` already checks we are in a valid state for that transition. You can remove this require.

**Comment by uivlis** [comment]
Yeah, when I run `yarn security` I get:
```
myth: error: unrecognized arguments: --truffle
```
@obernardovieira Am I missing something?

**Comment by obernardovieira** [comment]
`yarn security` does not work currently, because mythrill changed. I've got it working again a while ago, but it's not here. Don't use it for now. I'll open an issue to fix that. Sorry about that!

**Comment by obernardovieira** [comment]
Also, rebase from `dev` branch and fix conflicts when you see them.

**Comment by uivlis** [comment]
@obernardovieira Could not import `ganache-time-traveler` without `require`, it kept telling me that I don't have the `d.ts` file, even though it was there. Don't know what to do, any suggestions?

**Comment by obernardovieira** [comment]
> @obernardovieira Could not import `ganache-time-traveler` without `require`, it kept telling me that I don't have the `d.ts` file, even though it was there. Don't know what to do, any suggestions?

It seems to be an issue with typechain dependency. I tried to do as I usually do with other packages, and tslint says everything is ok, but when running `generate-ts` it fails. So I'll try to have something more specific and raise an issue at typechain repository.
Meanwhile, we can use it as it is. Just fix the conflicts and make CI work and I'll review it :+1:

**Pull request merged**